### ISSUE TITLE
Improve command descriptions

### ISSUE THREAD

**Issue by schlessera** [wp-cli/wp-cli#4302]
When you go to the command reference page at https://developer.wordpress.org/cli/commands/ , you can see rather useless descriptions such as these:

<img width="470" alt="screenshot 2017-08-16 07 37 55" src="https://user-images.githubusercontent.com/83631/29348692-dcda6c70-8255-11e7-8962-788c7509bc61.png">

We should go through the descriptions and rephrase them so that they actually add additional value, rather than stating the obvious.

**Comment by gitlost** [comment]
With reference to PR https://github.com/wp-cli/cache-command/pull/8, I like the style of the current `wp cli` short description:

> Review current WP-CLI info, check for updates, or see defined aliases.

so for instance for the cache command something like:

> Flush the WP Object Cache, manipulate its values, determine which cache is being used.

here leaving out "or" or "and" at the end as it's implied, which I'd also do in the `wp cli` case:

> Review current WP-CLI info, check for updates, see defined aliases.

**Comment by danielbachhuber** [comment]
cc @ako80218  ^

**Comment by hearvox** [comment]
We'll be proposing in this thread new descriptions for the top-level commands. If approved we'll fork/edit the file and submit a pull request. (The description grammar follows WP Code Ref pattern of starting with plural verb, e.g.: "Retrieves details...".) Here's the format of my suggestions:

URL (truth-src)
command | Current description.
Proposed:
First sentence of new description (<80 chars; becomes command post's excerpt).

Sometimes a second sentence to include a link (to Codex or Plugin/Theme Handbook reference).

**Comment by hearvox** [comment]
https://github.com/wp-cli/media-command/blob/master/src/Media_Command.php
wp media | Import new attachments or regenerate existing ones.
Proposed:
Imports files as attachments and regenerates thumbnails.

**Comment by hearvox** [comment]
https://github.com/wp-cli/entity-command/blob/master/src/Menu_Command.php
wp menu | List, create, assign, and delete menus.
Proposed:
Manages menus for the WordPress site.

Lists, creates, assigns, and deletes the active theme's [navigation menus](https://developer.wordpress.org/themes/functionality/navigation-menus/).

**Comment by hearvox** [comment]
(Note: This command has no description now: Don't think it has a truth-src file.
There's is for its subcommand:
https://github.com/wp-cli/entity-command/blob/master/src/Network_Meta_Command.php
Should we create Network_Command.php a file in the entity-command repo?)

wp network | (none)
Proposed:
Manages meta fields for multisite networks.

**Comment by hearvox** [comment]
https://github.com/wp-cli/entity-command/blob/master/src/Option_Command.php
wp option | Manage options.
Proposed:
Retrieves and sets site options, including plugin and WordPress settings.

See the [Plugin Settings API](https://developer.wordpress.org/plugins/settings/settings-api/) and the 
[Theme Options](https://developer.wordpress.org/themes/customize-api/) for more information on adding customized options.

**Comment by hearvox** [comment]
Or do you prefer we submit the proposed description chnages in the individual repos? If so, we will.

**Comment by schlessera** [comment]
@hearvox As the pull requests will need to go into the separate repos, having the issues in there as well with the discussion would improve linking.

Also, having a separate issue per proposal will also better allow us to discuss.

**Comment by ako80218** [comment]
@gitlost, @schlessera, @danielbachhuber I have received contradictory advice as to whether we should propose changes to command descriptions here or create a separate issue per proposal. @hearvox and I are happy to proceed either way but we need an agreement from you all as to which way to proceed. Thanks!

**Comment by schlessera** [comment]
@ako80218 Ah I see, sorry about that.

My thinking is that it is difficult to discuss the individual descriptions in here, as we'll have intertwining comments in a huge list in the end. That's why I proposed to create separate issues, because each issue will then collect one discussion thread. Also, it will allow us to close individual issues as soon as they are done with, instead of having one huge issue that goes on and on until the very last description has been completed.

@danielbachhuber, @gitlost What do you think? Let's agree on something first before proceeding.

**Comment by danielbachhuber** [comment]
> That's why I proposed to create separate issues, because each issue will then collect one discussion thread.

Given @hearvox has already gone and done this, I think we can continue with this plan.

**Comment by miya0001** [comment]
Some pull requests related to this are coming from some people.
I think it is difficult to judge whether it should be merged, because there will be various opinions.

So, I want to ping to @hearvox and @ako80218 from those pull-requests.
Could you check those pull-requests? 

If you feel good, please post `+1` or so. 😄 
We need consistency for those description, so it should be judged by same people.

Then we will review and merge. :)

How about this plan?

**Comment by hearvox** [comment]
miya0001,
@ako80218 and I wrote improved command descriptions together at WordCamp for Publishers two weeks ago. Since then we've tried in different ways to submit the descriptions, but the procedure has not been clear to us -- we've gotten conflicting instructions.

Our latest attempt was to:
* Submit the description as an issue in each repo.
* Wait for discussion.
* Revise, if there are suggestions.
* Submit a pull request for the change(s) in each repo..

@ako80218 and I are NOT submitting pull requests for the same commands: We've split them up: He's submitting some, me others. All the descriptions he and I are submitting as pull requests should be consistent: We wrote them together for consistency.

Are we still missing a step, or have we followed the required protocol to submit these simple improvements to the command descriptions?

**Comment by miya0001** [comment]
@hearvox 

> @ako80218 and I wrote improved command descriptions together at WordCamp for Publishers two weeks ago. Since then we've tried in different ways to submit the descriptions, but the procedure has not been clear to us -- we've gotten conflicting instructions.

Ah, I am sorry. I didn't know you are working for it since WordCamp for Publishers.

OK, I will check if PR will be merged on the following list:

* [ ] admin
* [ ] cache
* [ ] cap
* [ ] checksum
* [ ] cli
* [ ] comment
* [ ] config
* [ ] core
* [ ] cron
* [ ] db
* [ ] eval
* [ ] eval-file
* [ ] export
* [ ] help
* [ ] import
* [ ] language
* [ ] media
* [ ] menu
* [ ] network
* [ ] option
* [ ] package
* [ ] plugin
* [ ] post
* [ ] post-type
* [ ] rewrite
* [ ] role
* [ ] scaffold
* [ ] search-replace
* [ ] server
* [ ] shell
* [ ] sidebar
* [ ] site
* [ ] super-admin
* [ ] taxonomy
* [ ] term
* [ ] theme
* [ ] transient
* [ ] user
* [ ] vackup
* [ ] valet
* [ ] widget

Thanks!

**Comment by hearvox** [comment]
Thanks for the reply, @miya0001, and for approving the PRs. Now that we're sure we're following the correct procedure, @ako80218 and I will be submitting more suggested command descriptions as Issues, then as PRs. Appreciate your patience in ushering us thru this process.

**Comment by miya0001** [comment]
Thanks @hearvox , if you open a new pull-request, please link to this issue.
Committer will update the list as soon as possible. 😄

**Comment by hearvox** [comment]
@miya0001, thanks for adding links to this issue on my PRs. Once the first of those is merged (all are now Approved), I and @ako80218 will be sure we're doing everything right, then we'll submit more PRs with our already written command descriptions. FYI, mine so far:
https://github.com/wp-cli/rewrite-command/pull/9
https://github.com/wp-cli/package-command/pull/32
https://github.com/wp-cli/media-command/pull/40

**Comment by miya0001** [comment]
Thanks @hearvox 

Some people already open pull-request, please check existing PRs before open new PR. :)

**Comment by schlessera** [comment]
@hearvox As you said these have been discussed and decided in a team, would you mind providing the profile names of the people involved?
As someone else is already doing pull requests, I'd like to mention the contributor's profile names in each pull request.

**Comment by schlessera** [comment]
Ah, going through the above discussion, it seems it was only @hearvox and @ako80218 , is that correct?

**Comment by hearvox** [comment]
Correct, @schlessera:  @hearvox and @ako80218 

We wrote them together at a recent WordCamp. Got more PRs ready to submit once we're sure we've done everything right (i.e, after the first is merged).

**Comment by miya0001** [comment]
@schlessera @hearvox @ako80218 

I am thinking those PRs should be merged at once. :)
I want to do final check before merge them.
(I am sorry PR for `wp menu` already merged in my mistake.)

**Comment by hearvox** [comment]
Added 2 more PRs, both with multiple file changes:
https://github.com/wp-cli/media-command/pull/40
https://github.com/wp-cli/extension-command/pull/39

Still waiting/hoping for our first merge. If we've done something incorrectly in submitting these new command description, please tell us. (Or maybe this long a post-approval wait for merge is normal, in which case: I just gotta learn more patience.)

**Comment by miya0001** [comment]
I feel uncomfortable at the position of the line break when it has long description.
I want to hear opinion from @wp-cli/committers .

Example:
![](https://www.evernote.com/l/ABUl38sLavhHHYnWWcyuDK5eKwbh6xwYgjMB/image.png)

**Comment by hearvox** [comment]
Long and multiple sentences in the descriptions are a problem because forced line breaks read OK in the code, for example, of this existing description:
https://github.com/hearvox/shell-command/edit/master/src/Shell_Command.php

But are grammatically wrong, bad for a11y, and look funny when imported into the Handbook:
https://developer.wordpress.org/cli/commands/shell/

**Comment by gitlost** [comment]
> I feel uncomfortable at the position of the line break when it has long description.

@miya0001 the word wrapping for the command line help deals poorly with long non-spaced runs, in particular with links such as `Capabilities](https://codex.wordpress.org/Roles_and_Capabilities)` - seeing as they're becoming common in the new descriptions it should deal with them better. I'll open an issue for this.

> Long and multiple sentences in the descriptions are a problem because forced line breaks...look funny when imported into the Handbook

@hearvox again this looks to me that this should be programmatically fixed when generating the handbook. I'll open an issue for this also.

I suppose it's too late but I'm not a fan of using the third person singular instead of the imperative in the descriptions. These are commands and
```
Add, remove, and list capabilities of a user role.
```
reads better to me than the passive
```
Adds, removes, and lists capabilities of a user role.
```
If this was done in analogy to function descriptions then I think it's misapplied...(but it's not a big deal, and it's too late anyway!).

**Comment by miya0001** [comment]
@hearvox 

Do you have list of descriptions? Can you share it?
I am thinking we need to review whole descriptions.

@gitlost 

Sorry, I guess I shouldn't have merged some changes yet.
But it is not too late, we have a couple of month until next release. 😄 

I want to solve the word-wrapping problem in long description in a few weeks or so. #4336 
But I am not sure about the third-person-singular problem. We need more opinion for it.

**Comment by gitlost** [comment]
Well what's been merged is better than what was there before so that's fine I think, and I don't want to derail these improvements or bog them down in small details. So I think leaving things progress as now (without an overall review) may be best, and as you say it's not too late and we can always iterate.

> I want to solve the word-wrapping problem in long description in a few weeks or so

We will!

> But I am not sure about the third-person-singular problem.

Yes, that's definitively just my take, and others may differ.

**Comment by schlessera** [comment]
@miya0001, @gitlost:

Okay, let's make progress now and merge the descriptions that are not controversial. I agree that they are better than what we have so far, and we can still tweak them later on.

Here are the checks I want us to go through for each description:

1. Make sure the language is consistent with the new format (which seems to generally be the case).
2. Make sure all main functionalities are covered (which might need to be done at differing granularities to keep the length in check).
3. Apply the PR and make sure the description displays properly (mainly because of the length).
4. For all merged PRs, have the two original authors of the description (@hearvox & @ako80218) be part of the issue's props, even if they didn't create the actual PR.

If you see a description PR that matches the above, feel free to merge. Once we've merged them all, we'll have the actual command reference page as an overview to discuss any final tweaking.

**Comment by hearvox** [comment]
> third person singular instead of the imperative in the descriptions

@gitlost, We started command descriptions with a third-person-singular verb to match most common case in the Code Reference resource, as in these recent functions: 
https://developer.wordpress.org/reference/since/4.8.0/

Aiming for consistency between WP dev resources would be a good target. But maybe more important is consistency within the same resource. The WP-CLI Commands reference varies grammatically tho many sub-command descriptions do start with the imperative:
https://developer.wordpress.org/cli/commands/cache/

If that's the choice of the WP-CLI crew, maybe state it somewhere, for when stray Docs folk like myself come wondering thru. Maybe that somewhere could state other WP-CLI docs preferences myself and @ako80218 stumbled upon or were told along the way, like:
* The source of truth for command descriptions are in each repo's /src/ dir.
* The command description's first sentence become the `post_excerpt` in the WP-CLI Commands reference.
* Try to keep first-sentence line lengths <81 chars.
* Avoid unneeded "the", "of" and "and" words when possible.
 
Had we known the above earlier we would have saved ourselves many hours and rewrites. I'd be happy to write these up into a file somewhere, if y'all think it's worth it and tell me what file.

**Comment by schlessera** [comment]
@hearvox Here's what we currently have in terms of how to document commands: https://make.wordpress.org/cli/handbook/commands-cookbook/#optional-registration-arguments

**Comment by gitlost** [comment]
> If that's the choice of the WP-CLI crew, maybe state it somewhere, for when stray Docs folk like myself come wondering thru.

There's no us and them here. It was just a suggestion, by me alone, as stated above.

I find the tone of this conversation tiresome, and won't be participating further.

**Comment by hearvox** [comment]
@gitlost, Hope I didn't offend. I certainly didn't mean to. You brought up a great point about the "person" of the starting verb in these command descriptions. The WP-CLI Command reference and the Codex/Code Reference you both the 3rd-singular and the imperative. Both make sense, and whichever way y'all want to go is the way we'll write 'em. But for now, unless we hear different, we'll stick w/ 3rd-singular.

**Comment by gitlost** [comment]
@hearvox thanks for responding, no offence taken.

**Comment by danielbachhuber** [comment]
Thanks again for your persistence on this issue, @hearvox. It's been a helpful exercise in reviewing these descriptions.

**Comment by danielbachhuber** [comment]
Pretty happy with these for now. We can continue to work on them in the future.
